api/envoy/config/core/v3/health_check.proto

```diff
@@ -145,6 +145,10 @@ message HealthCheck {
     // <envoy_v3_api_msg_type.matcher.v3.StringMatcher>`. See the :ref:`architecture overview
     // <arch_overview_health_checking_identity>` for more information.
     type.matcher.v3.StringMatcher service_name_matcher = 11;
+
+    // An optional string to match anywhere in the first 1024 bytes of the response body.
+    // If it is set, both the expected response check and status code determine the health check.
```

Comment:
Interesting. I think to make this generic, we can add BinaryMatcher similar to StringMatcher https://github.com/envoyproxy/envoy/blob/main/api/envoy/type/matcher/v3/string.proto 

We can also have `one of` here of BinaryMatcher and StringMatcher for convenience.

Reply:
Thank you! Can the generic job be a new PR from Api's point of view?

Reply:
It will require deprecation of the existing field and writing code that takes care of compatibility. It is more time efficient to do it right from the start, instead of pushing expected API changes to a later date.

Reply:
Got it, let me look into it. Thank you!

Reply:
Sorry, I'm confused. Why doesn't this look more like the `Payload` message already used by `TcpHealthCheck`? That seems the place to try and harmonize.

Reply:
We need to choose one of [health checkers](https://github.com/envoyproxy/envoy/blob/1cf5603dc5239c92e5bc38ef321f59ccf6eabc6e/api/envoy/config/core/v3/health_check.proto#L288). here we mainly want to support HTTP for Google Cloud [feature](https://cloud.google.com/load-balancing/docs/health-check-concepts#criteria-protocol-http). 

Added a BinaryMacher, and make them generic.

Reply:
Not sure I follow - why can't we teach the `Payload` message, which is already in the business of binary matching, to do a better job for both TCP and HTTP body health check? This is the same problem I think and we should have similar levels of capability in both place.

Reply:
Good point, thank you for pointing it out! 

I think we can reuse the un-implemented [Payload](https://github.com/envoyproxy/envoy/blob/1cf5603dc5239c92e5bc38ef321f59ccf6eabc6e/api/envoy/config/core/v3/health_check.proto#L106-L107) in HTTP health check proto. And then create our own `HttpHealthCheckMatcher` class, and probably in the future we can add the `match_pattern` proto filed in [message HttpHealthCheck]() and TcpHealthCheck proto to extend.  The default behavior for HTTP health check is the contains pattern, and the default behavior for TCP is the Mongo HEX pattern for now.

We probably can make this [TcpHealthCheckMatcher](https://github.com/envoyproxy/envoy/blob/23ff23d0954eb9e6e4ed7ed60ec0e0d05174f6da/source/common/upstream/health_checker_impl.h#L194-L244) generic. Tcp health check payload match is using MongoDB health check and is only using HEX string, and doesn't implement bytes. 

How do you think? Thank you!

Reply:
I think the idea of standardizing on Payload makes sense, and then we can extend Payload for any more sophisticated matching, on string or bytes.

Reply:
Hi all, can you review again? I used Payload api, and the match pattern can be extendable in future. Thank you!

Reply:
Yes, makes sense now just one docs comment, thanks.